Add nullbits nibble

jaygreco opened the pull request:
## Description

Add `nullbitsco` folder and `nibble` keyboard.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* None

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

tzarc commented on keyboards/nullbitsco/nibble/big_led.h:
```diff
@@ -0,0 +1,67 @@
+        break;
+
+        default:
+        break;
+    }
+}
+
+void set_big_LED_g(uint8_t mode) {
+    switch(mode) {
+        case LED_ON:
+            setPinOutput(BIG_LED_G_PIN);
+            writePin(BIG_LED_G_PIN, GPIO_STATE_HIGH);
+        break;
+
+        case LED_OFF:
+            setPinOutput(BIG_LED_G_PIN);
+            writePin(BIG_LED_G_PIN, GPIO_STATE_LOW);
+        break;
+
+        default:
+        break;
+    }
+}
+
+void set_big_LED_b(uint8_t mode) {
+    switch(mode) {
+        case LED_ON:
+            setPinOutput(BIG_LED_B_PIN);
+            writePin(BIG_LED_B_PIN, GPIO_STATE_HIGH);
+        break;
+
+        case LED_OFF:
+            setPinOutput(BIG_LED_B_PIN);
+            writePin(BIG_LED_B_PIN, GPIO_STATE_LOW);
+        break;
+
+        default:
+        break;
+    }
+}
```
You'd be best moving these to a `big_led.c` file, and leaving the function declarations in the header.

tzarc commented on keyboards/nullbitsco/nibble/matrix.c:
```diff
@@ -0,0 +1,167 @@
+/* Copyright 2018 Jack Humbert
```
This entire file can probably be significantly simplified by swapping to `CUSTOM_MATRIX = lite` in your `rules.mk`.
All things such as debounce and the like are abstracted away, and the implementation just needs to provide the raw gathering of the matrix data.

You'd only need to provide:
```c
void matrix_init_custom(void) { /* ... */ }
bool matrix_scan_custom(matrix_row_t current_matrix[]) {
    bool has_changed = false;
    /* ... */
    return has_changed;
}
```

tzarc commented on keyboards/nullbitsco/nibble/nibble.c:
```diff
@@ -0,0 +1,85 @@
+}
+#endif
+
+void matrix_init_kb(void) {
+    matrix_init_user();
+}
+
+void matrix_scan_kb(void) {
+    matrix_scan_user();
+}
+
+bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
+    return process_record_user(keycode, record);
+}
+
```
No need for these, they do this already.
```suggestion
```

tzarc replied:
Still needs removal.

tzarc commented on keyboards/nullbitsco/nibble/README.md:
```diff
@@ -0,0 +1,15 @@
+
+Make example for this keyboard (after setting up your build environment):
+
+    make nullbitsco/nibble
```
```suggestion
    make nullbitsco/nibble:default
```

jaygreco commented:
Hi @tzarc! Thanks for the feedback. I'll get working on making the changes you suggested on my local branch. Good info on the CUSTOM_MATRIX = lite. I wasn't aware of that option. I'll await more detailed feedback from you or the other collaborators before updating the PR further.

fauxpark commented on keyboards/nullbitsco/nibble/keymaps/via/nibble.json:
```diff
@@ -0,0 +1,29 @@
+{
```
This file belongs in the VIA repo, not here.

fauxpark commented on keyboards/nullbitsco/nibble/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,81 @@
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+  [_MA] = LAYOUT(
+               KC_ESC,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_TILD, \
+    KC_F13,    KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS, KC_DEL,  \
+    KC_F14,    KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_ENT,  KC_PGUP, \
+    KC_F15,    KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSFT,          KC_UP,   KC_PGDN, \
+    KC_F16,    KC_LCTL, KC_LGUI, KC_LALT,                   KC_SPC,                    MO(_FN), KC_RALT, KC_RCTL, KC_LEFT,          KC_DOWN, KC_RGHT  \
+  ),
+  [_FN] = LAYOUT(
+               RESET,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,  KC_HOME,  KC_INS, \
+    RGB_TOG, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, \
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, \
+    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______, \
+    _______, _______, _______, _______,                   _______,                   _______, _______, _______, KC_MPRV,          KC_MPLY, KC_MNXT  \
```
```suggestion
               KC_ESC,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_TILD,
    KC_F13,    KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS, KC_DEL,
    KC_F14,    KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_ENT,  KC_PGUP,
    KC_F15,    KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSFT,          KC_UP,   KC_PGDN,
    KC_F16,    KC_LCTL, KC_LGUI, KC_LALT,                   KC_SPC,                    MO(_FN), KC_RALT, KC_RCTL, KC_LEFT,          KC_DOWN, KC_RGHT
  ),
  [_FN] = LAYOUT(
               RESET,   KC_F1,   KC_F2,   KC_F3,   KC_F4,   KC_F5,   KC_F6,   KC_F7,   KC_F8,   KC_F9,   KC_F10,  KC_F11,  KC_F12,  KC_HOME,  KC_INS,
    RGB_TOG, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,
    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______,
    _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______, _______,          _______, _______,
    _______, _______, _______, _______,                   _______,                   _______, _______, _______, KC_MPRV,          KC_MPLY, KC_MNXT
```

fauxpark commented on keyboards/nullbitsco/nibble/rules.mk:
```diff
@@ -0,0 +1,73 @@
+#     automatically to create a 32-bit value in your source code.
+#
+#     This will be an integer division of F_USB below, as it is sourced by
+#     F_USB after it has run through any CPU prescalers. Note that this value
+#     does not *change* the processor frequency - it should merely be updated to
+#     reflect the processor speed set externally so that the code can use accurate
+#     software delays.
+F_CPU = 16000000
+
+
+#
+# LUFA specific
+#
+# Target architecture (see library "Board Types" documentation).
+ARCH = AVR8
+
+# Input clock frequency.
+#     This will define a symbol, F_USB, in all source code files equal to the
+#     input clock frequency (before any prescaling is performed) in Hz. This value may
+#     differ from F_CPU if prescaling is used on the latter, and is required as the
+#     raw input clock is fed directly to the PLL sections of the AVR for high speed
+#     clock generation for the USB and other AVR subsections. Do NOT tack on a 'UL'
+#     at the end, this will be done automatically to create a 32-bit value in your
+#     source code.
+#
+#     If no clock division is performed on the input clock inside the AVR (via the
+#     CPU clock adjust registers or the clock division fuses), this will be equal to F_CPU.
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+
+# Boot Section Size in *bytes*
+#   Teensy halfKay   512
+#   Teensy++ halfKay 1024
+#   Atmel DFU loader 4096
+#   LUFA bootloader  4096
+#   USBaspLoader     2048
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
```
```suggestion
# Bootloader selection
#   Teensy       halfkay
#   Pro Micro    caterina
#   Atmel DFU    atmel-dfu
#   LUFA DFU     lufa-dfu
#   QMK DFU      qmk-dfu
#   ATmega32A    bootloadHID
#   ATmega328P   USBasp
BOOTLOADER = atmel-dfu
```

fauxpark commented on keyboards/nullbitsco/nibble/rules.mk:
```diff
@@ -0,0 +1,73 @@
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no      	# Virtual DIP switch configuration(+1000)
+MOUSEKEY_ENABLE = no       	# Mouse keys(+4700)
+EXTRAKEY_ENABLE = yes       # Audio control and System control(+450)
+# CONSOLE_ENABLE = yes        # Console for debug(+400)
+COMMAND_ENABLE = no         # Commands for debug and configuration
+CUSTOM_MATRIX = yes         # Custom matrix file
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = no     	# Enable keyboard backlight functionality
+MIDI_ENABLE = no            # MIDI support (+2400 to 4200, depending on config)
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+ENCODER_ENABLE = yes		# Enable encoder
+RGBLIGHT_ENABLE = yes
+LTO_ENABLE = yes
+# OLED_DRIVER_ENABLE = yes
```
```suggestion
BOOTMAGIC_ENABLE = no      	# Virtual DIP switch configuration
MOUSEKEY_ENABLE = no       	# Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no     	# Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
MIDI_ENABLE = no            # MIDI support
BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
AUDIO_ENABLE = no           # Audio output on port C6
FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
ENCODER_ENABLE yes
LTO_ENABLE = yes
CUSTOM_MATRIX = yes

# Project specific files
SRC += matrix.c
```

fauxpark commented on keyboards/nullbitsco/nibble/rules.mk:
```diff
@@ -0,0 +1,73 @@
+# Project specific files
+SRC += matrix.c
+
```
```suggestion
```

drashna commented on keyboards/nullbitsco/nibble/config.h:
```diff
@@ -0,0 +1,61 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef CONFIG_H
+#define CONFIG_H
```
```suggestion
#pragma once
```

drashna commented on keyboards/nullbitsco/nibble/config.h:
```diff
@@ -0,0 +1,61 @@
+#define ENCODERS_PAD_A { B5 }
+#define ENCODERS_PAD_B { B4 }
+
+#endif
```
```suggestion
```

jaygreco commented:
Thanks for the feedback everyone. I'm going to be marking these as resolved as I take care of them locally. I'm going to consolidate all of these into one commit.

drashna commented on keyboards/nullbitsco/nibble/serial.h:
```diff
@@ -0,0 +1,91 @@
+#pragma once
```
We have the `/drivers/avr/serial.c`, and depending on what you're trying to do, you may be able to use that.

fauxpark replied:
That's software serial. It would probably be better to use `tmk_core/common/uart.c` or `tmk_core/protocol/serial_uart.c`.

jaygreco replied:
Thanks for the heads up! I didn't realize those changes had been included in `tmk_core/common/uart.c`. I'll refactor to leverage that code instead of my own.

drashna commented on keyboards/nullbitsco/nibble/README.md:
```diff
@@ -0,0 +1,15 @@
+A unique, tweakable 65% keyboard kit built by nullbits. [More info at nullbits.co](https://nullbits.co/nibble/)
+
+* Keyboard Maintainer: [Jay Greco](https://github.com/jaygreco)
+* Hardware Supported: NIBBLE Rev1, [Bit-C](https://nullbits.co/bit-c/) Pro Micro with USB-C.
```
Rather, this is compatible with **_any_** Pro Micro compatible controller?

drashna commented on keyboards/nullbitsco/nibble/config.h:
```diff
@@ -0,0 +1,63 @@
+
+/* Used to set master for remote KB if VUSB detect doesn't work. */
+// #define KEYBOARD_MASTER
+#define NO_USB_STARTUP_CHECK
```
Is this needed?

nblyumberg replied:
Might be needed for MacOS wake from sleep? I vaguely recall reading that somewhere.

jaygreco replied:
@nblyumberg I apologize, I let this one slip through the cracks. You're exactly right; I followed the suggestion [here](https://github.com/qmk/qmk_firmware/issues/5585#issuecomment-491255880). I followed the rest of the thread but was still seeing the issue with the latest QMK.

drashna replied:
To be honest/blunt, any info from that thread wouldn't apply to this board, as the issue there is with chibiOS, and therefore ARM boards. 

This board appears to be an AVR board, and not running any chibiOS related code.

jaygreco replied:
@drashna Got it, I skimmed that thread and didn't realize the HS60 was an ARM board. It is an issue I've seen in the wild (including repro on my own keeb) and confirmed it was resolved by adding `NO_USB_STARTUP_CHECK`. I'd prefer it stay in since most users are just flashing via QMK configuratior or toolbox and aren't building a binary themselves.

If it's an absolute no-go and removing it will get this PR through, then so be it. I'll add a note to the readme either way.

drashna commented on keyboards/nullbitsco/nibble/config.h:
```diff
@@ -0,0 +1,63 @@
+#define MATRIX_COLS 16
+#define MATRIX_MUX_COLS 4
+
+/* prevent stuck modifiers */
+#define PREVENT_STUCK_MODIFIERS
```
This is no longer used, as it is the default behavior.  

```suggestion
```

drashna commented on keyboards/nullbitsco/nibble/config.h:
```diff
@@ -0,0 +1,63 @@
+/* Optional SMT LED pins */
+#define RGB_DI_PIN E6
+#define RGBLED_NUM 10
+#define RGBLIGHT_LIMIT_VAL 255
```
This isn't needed, if it's set to 255, as that is the default.

drashna commented on keyboards/nullbitsco/nibble/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,151 @@
+#define _VIA2     2
+#define _VIA3     3
+
+#include "quantum.h"
```
Not needed here, included by default. 
```suggestion
```

drashna commented on keyboards/nullbitsco/nibble/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,151 @@
+#define _VIA3     3
+
+#include "quantum.h"
+#include "remote_kb.h"
```
```suggestion
```

drashna commented on keyboards/nullbitsco/nibble/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,81 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+#include "remote_kb.h" 
```
This should not be included here.  If this is needed for the keymap to function properly, it needs to be included in the keyboard's header file, instead. 

```suggestion
```

drashna commented on keyboards/nullbitsco/nibble/nibble.h:
```diff
@@ -0,0 +1,36 @@
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#ifndef NIBBLE_H
+#define NIBBLE_H
```
```suggestion
#pragma once

```

drashna commented on keyboards/nullbitsco/nibble/nibble.h:
```diff
@@ -0,0 +1,36 @@
+#define NIBBLE_H
+#define ___ KC_NO
+
+#include "quantum.h"
```
```suggestion
#include "quantum.h"
#include "remote_kb.h"
```

drashna commented on keyboards/nullbitsco/nibble/rules.mk:
```diff
@@ -0,0 +1,75 @@
+# MCU name
+MCU = atmega32u4
+
+# Processor frequency.
+#     This will define a symbol, F_CPU, in all source code files equal to the
+#     processor frequency in Hz. You can then use this symbol in your source code to
+#     calculate timings. Do NOT tack on a 'UL' at the end, this will be done
+#     automatically to create a 32-bit value in your source code.
+#
+#     This will be an integer division of F_USB below, as it is sourced by
+#     F_USB after it has run through any CPU prescalers. Note that this value
+#     does not *change* the processor frequency - it should merely be updated to
+#     reflect the processor speed set externally so that the code can use accurate
+#     software delays.
+F_CPU = 16000000
+
+
+#
+# LUFA specific
+#
+# Target architecture (see library "Board Types" documentation).
+ARCH = AVR8
+
+# Input clock frequency.
+#     This will define a symbol, F_USB, in all source code files equal to the
+#     input clock frequency (before any prescaling is performed) in Hz. This value may
+#     differ from F_CPU if prescaling is used on the latter, and is required as the
+#     raw input clock is fed directly to the PLL sections of the AVR for high speed
+#     clock generation for the USB and other AVR subsections. Do NOT tack on a 'UL'
+#     at the end, this will be done automatically to create a 32-bit value in your
+#     source code.
+#
+#     If no clock division is performed on the input clock inside the AVR (via the
+#     CPU clock adjust registers or the clock division fuses), this will be equal to F_CPU.
+F_USB = $(F_CPU)
```
All of this isn't needed, as it's set by default. 
```suggestion
```

jaygreco commented:
Hi all, don't want to be a bother, but I wanted to check in and see what is needed to get this request approved. I've addressed the feedback from the reviews and added the requested changes into the codebase. Please let me know, thanks! 
@drashna @tzarc @fauxpark

jaygreco commented:
Looks like the linter is failing after I made a recent commit adding ISO compatibility. 
@drashna, @tzarc, or @fauxpark, any insight into why? I really want to push this PR through, I have a VIA PR waiting on it.
Do let me know what's needed, thanks!

tzarc commented:
Yeah looks like something weird with CI, we'll investigate but it doesn't seem like it's your end.
I'll try to get to this PR at some point tonight, but I can't guarantee.

jaygreco commented:
Thanks @tzarc! Good to know. I've made all of the requested updates up to this point, so hopefully reviewing's light work. Thanks for the super-fast response!

tzarc commented on keyboards/nullbitsco/nibble/keymaps/iso/keymap.c:
```diff
@@ -0,0 +1,80 @@
+return true;
+}
+
+void matrix_init_user(void) {
+  matrix_init_remote_kb();
+}
+
+void matrix_scan_user(void) {
+  matrix_scan_remote_kb();
+}
```
Same here.

tzarc commented on keyboards/nullbitsco/nibble/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,148 @@
+return true;
+}
+
+void matrix_init_user(void) {
+  matrix_init_remote_kb();
+}
+
+void matrix_scan_user(void) {
+  matrix_scan_remote_kb();
```
...and here.

tzarc commented on keyboards/nullbitsco/nibble/config.h:
```diff
@@ -0,0 +1,59 @@
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    nullbits
+#define PRODUCT         NIBBLE
+#define DESCRIPTION     QMK keyboard firmware
```
```suggestion
```

tzarc commented on keyboards/nullbitsco/nibble/nibble.c:
```diff
@@ -0,0 +1,85 @@
+    bool alt = get_mods() & MOD_MASK_ALT;
+    bool ctrl = get_mods() & MOD_MASK_CTRL;
+
+    if (clockwise) {
+        if (alt) {
+            rgblight_increase_hue();
+        } else if (ctrl) {
+            rgblight_increase_val();
+        } else if (shift) {
+            rgblight_increase_sat();
+        } else {
+            rgblight_step();
+        }
+
+  } else {
+      if (alt) {
+            rgblight_decrease_hue();
+        } else if (ctrl) {
+            rgblight_decrease_val();
+        } else if (shift) {
+            rgblight_decrease_sat();
+        } else {
+            rgblight_step_reverse();
+        }
+    } 
+}
+
+void encoder_update_kb(uint8_t index, bool clockwise) {
+  if (layer_state_is(1)) {
+    //change RGB settings
+    change_RGB(clockwise);
+  }
+  else {
+    if (clockwise) {
+      tap_code(KC_VOLU);
+  } else {
+      tap_code(KC_VOLD);
+    }  
+  }
+}
```
Let's get these into a keymap instead.

tzarc commented on keyboards/nullbitsco/nibble/nibble.c:
```diff
@@ -0,0 +1,85 @@
+#include QMK_KEYBOARD_H
+#include "bitc_led.h"
+
+#ifdef VIA_ENABLE
+#include "via_extras.h"
+
+#else
```
Can move this directly into the via keymap.

tzarc commented on keyboards/nullbitsco/nibble/nibble_encoder.h:
```diff
@@ -0,0 +1,140 @@
+#include QMK_KEYBOARD_H
+#include "tmk_core/common/eeprom.h"
```
Unsure if this is required, but......
```suggestion
#include "eeprom.h"
```

tzarc commented on keyboards/nullbitsco/nibble/nibble_encoder.h:
```diff
@@ -0,0 +1,140 @@
+    case ENC_MODE_CUSTOM2:
+      mapped_code = retrieve_custom_encoder_config(2, ENC_CUSTOM_CW);
+      break;
+#endif
+  }
+  return mapped_code;
+}
+
+uint16_t handle_encoder_ccw(void) {
+	dprintf("Encoder mode: %u\n", encoder_mode);
+  uint16_t mapped_code = 0;
+  switch(encoder_mode){
+    default:
+    	break;
+    case ENC_MODE_VOLUME:
+      mapped_code = KC_VOLD;
+      break;
+    case ENC_MODE_MEDIA:
+      mapped_code = KC_MEDIA_PREV_TRACK;
+      break;
+    case ENC_MODE_SCROLL:
+      mapped_code = KC_WH_U;
+      break;
+    case ENC_MODE_BACKLIGHT:
+      mapped_code = RGB_VAD;
+      break;
+#ifdef DYNAMIC_KEYMAP_ENABLE
+    case ENC_MODE_CUSTOM0:
+      mapped_code = retrieve_custom_encoder_config(0, ENC_CUSTOM_CCW);
+      break;
+    case ENC_MODE_CUSTOM1:
+      mapped_code = retrieve_custom_encoder_config(1, ENC_CUSTOM_CCW);
+      break;
+    case ENC_MODE_CUSTOM2:
+      mapped_code = retrieve_custom_encoder_config(2, ENC_CUSTOM_CCW);
+      break;
+#endif
+  }
+  return mapped_code;
+}
```
Change these to declarations, and move the implementations into a source file.

tzarc commented on keyboards/nullbitsco/nibble/remote_kb.c:
```diff
@@ -0,0 +1,128 @@
+#include "remote_kb.h"
+
```
Lets have some documentation here -- I had no idea what `remote_kb` is.
Perhaps rename the file to something that matches the actual hardware, along with the exported APIs?

tzarc commented on keyboards/nullbitsco/nibble/keymaps/iso/keymap.c:
```diff
@@ -0,0 +1,80 @@
+/* Copyright 2018 Jack Humbert
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
```
I'd say we remove this entire keymap, given that "LAYOUT_iso" is a valid option in general.

tzarc commented on keyboards/nullbitsco/nibble/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,148 @@
+#include QMK_KEYBOARD_H
```
~~nibble_encoder.h, via_extras.h needs to be moved to "bells-and-whistles" style keymaps -- there's extra stuff over and above what's needed to get a board up and running, stuff like discord mute and the like aren't going to apply to everyone~~ -- discussed on Discord directly, and I don't think QMK has a proper solution for this, but we need to investigate how to solve this going forward. Bending the requirements in the PR checklist somewhat to allow this through -- usability of the kb for users was deemed more of a benefit than requiring "bells-and-whistles" keymaps for both `default` and `via` keymaps.

tzarc commented on keyboards/nullbitsco/nibble/rules.mk:
```diff
@@ -0,0 +1,42 @@
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
```suggestion
```

tzarc reviewed (changes requested):
Also, as discussed on Discord:
- need license headers on all source files

tzarc commented on keyboards/nullbitsco/nibble/keymaps/via/nibble_encoder.c:
```diff
@@ -0,0 +1,133 @@
+}
+
+void pre_encoder_mode_change(void) {
+	dprintf("Changing encoder mode.\n", encoder_mode);
```
```suggestion
	dprintf("Changing encoder mode: %u\n", encoder_mode);
```

tzarc reviewed (approved):
Just something silly.

drashna commented on keyboards/nullbitsco/nibble/nibble.c:
```diff
@@ -0,0 +1,25 @@
+#include "bitc_led.h"
+
+// Use Bit-C LED to show CAPS LOCK status
+void led_set_kb(uint8_t usb_led) {
+    if (usb_led & (1<<USB_LED_CAPS_LOCK))
+        set_bitc_LED(LED_DIM);
+    else
+        set_bitc_LED(LED_OFF);
+}
```
```suggestion
bool led_update_kb(led_t led_state) {
    bool res = led_update_user(led_state);
    if (res) {
        set_bitc_LED(led_state.capslock ? LED_DIM : LED_OFF);
    }
    return res;
}
```

drashna commented on keyboards/nullbitsco/nibble/rules.mk:
```diff
@@ -0,0 +1,35 @@
+CUSTOM_MATRIX = lite        # Lite custom matrix 
+
+# Project specific files
+SRC += matrix.c
+SRC += bitc_led.c
+SRC += big_led.c
+SRC += remote_kb.c
+SRC += tmk_core/common/uart.c
```
Just a nit pick: 
```suggestion
SRC += matrix.c \
       bitc_led.c \
       big_led.c \
       remote_kb.c \
       tmk_core/common/uart.c
```
